src/history/hs_rec.c

```diff
@@ -197,6 +197,24 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
         cursor->set_key(cursor, 3, btree->id, key, tw->start_ts + 1);
         WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, cursor), true);
     }
+
+    /*
+     * It is possible to insert a globally visible update into the history store with larger
+     * timestamps ahead of it. An example would be a mixed-mode update getting moved to the history
+     * store. This scenario can avoid detection earlier in reconciliation and result in an EBUSY
+     * being returned as it detects out-of-order timestamps. To prevent this we allow globally
+     * visible updates to fix history store content even if eviction is running concurrently with a
+     * checkpoint.
+     *
+     * This is safe because global visibility considers the checkpoint transaction id and timestamp
+     * while it is running, i.e. if the update is globally visible to eviction it will be globally
+     * visible to checkpoint and the modifications it makes to the history store will be the same as
+     * what checkpoint would've done.
+     */
+    if (error_on_ooo_ts && __wt_txn_tw_start_visible_all(session, tw)) {
```

Comment:
What about the stop timestamp?

Reply:
I think a out of order stop timestamp can trigger this as well.

Reply:
Would we ever move a globally visible stop timestamp to the history store?

Reply:
```
            /*
             * No need to continue if we see the first self contained value after the first globally
             * visible value.
             */
            if (first_globally_visible_upd != NULL && WT_UPDATE_DATA_VALUE(upd))
                break;
```
Yes, it's possible if a globally visible tombstone is newer to a modify.
Like, T -> M.

Reply:
Okay so you think this check should be `if (error_on_ooo_ts && (__wt_txn_tw_start_visible_all(session, wt) || __wt_txn_tw_stop_visible_all(session, wt)))`?

Reply:
yes, I think so. We should check stop before the start actually.

Reply:
Actually, if stop is globally visible, start must be also globally visible. So checking start is enough here I think.

Reply:
Actually if stop is globally visible, start must be globally visible as well. So your original check is enough I think.

Reply:
Fixed in: 1438d09

Reply:
The implication would be that the stop is out of order with the start for the stop to be globally visible and the start to not be but that would be fixed in the `hs_insert_updates` code path, correct?

Reply:
Yes, that's correct.

Reply:
I missed this discussion, we don't insert the update if the stop is globally visible as no one needs to see this data. The out of order stop should have been already fixed so that both start and stop should be the same values with timestamp?

it is not a problem in checking the stop also, I feel it is just a redundant check.

Reply:
Reverted the stop change.